src/OpenTelemetry/Trace/ExceptionProcessor.cs

```diff
@@ -0,0 +1,53 @@
+// <copyright file="ExceptionProcessor.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace OpenTelemetry.Trace
+{
+    internal class ExceptionProcessor : BaseProcessor<Activity>
+    {
+        private readonly Func<IntPtr> fnGetExceptionPointers;
+
+        public ExceptionProcessor()
+        {
+            try
+            {
+                var flags = BindingFlags.Static | BindingFlags.Public;
+                var method = typeof(Marshal).GetMethod("GetExceptionPointers", flags, null, new Type[] { }, null);
+                var lambda = Expression.Lambda<Func<IntPtr>>(Expression.Call(method));
+                this.fnGetExceptionPointers = lambda.Compile();
+            }
+            catch (Exception ex)
+            {
+                throw new NotSupportedException("System.Runtime.InteropServices.Marshal.GetExceptionPointers is not supported.", ex);
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnEnd(Activity activity)
```

Comment:
It maybe rare, but do we need ignore the exception if it is there in the call of `OnStart`, and only set the error status if the the exception pointer in `OnEnd` is different than the one in `OnStart`? Probably this is not 100% accurate if GC moves the exception object.

Reply:
I'm not sure if I understand this, would you elaborate?

Reply:
I meant for the check `his.fnGetExceptionPointers() != IntPtr.Zero` in `OnEnd()`, we assume the exception only happens after the call to `OnStart()` for the same processor, right? If so, we need to check the exception didn't happen in the call of `OnStart()`.

Reply:
I might be confused - if exception happens before `OnStart`, is it possible for `OnEnd` to get called?

Reply:
In the sample code like below, or catcher handler calls any function which starts a new activity, do we expect the new `Activity` to record this exception? I think the new activity doesn't need to record the exception because it happened before the creation of the `activity` but the current implementation seems record it.

```csharp
try {
    throw new Exception();
} catch (Exception e) {
    var activity = activitySource.StartActivity("ActivityForException");
}
```

Reply:
Now I see what you meant, that's a good catch.

Reply:
I've put a fix on this.
There are still corner cases which I guess can be considered as "too rare"?

```csharp
Activity activity = null;
try
{
}
catch (Exception)
{
    activity = activitySource.StartActivity("foobar");
}
finally
{
    activity?.Dispose(); // here we will have a tag with name `otel.exception_pointers`, we could address this as well but the perf penalty would be high (e.g. we can always check for `otel.exception_pointers` and remove it)
}
```